Block old legacy clients

LukasReschke opened the pull request:
This Pull Request introduces a SabreDAV plugin that will block all older clients than 1.6.1 to connect and sync with the ownCloud instance.

This has multiple reasons:
1. Old ownCloud client versions before 1.6.0 are not properly working with sticky cookies for load balancers and thus generating sessions en masse
2. Old ownCloud client versions tend to be horrible buggy

In some cases we had in 80minutes about 10'000 sessions created by a single user. While this change set does not really "fix" the problem as 3rdparty legacy clients are affected as well, it is a good work-around and hopefully should force users to update their client

@danimo @dragotin @DeepDiver1975 @karlitschek @butonic … and whoever might feel involved :see_no_evil:

LukasReschke committed:
Block old legacy clients

This Pull Request introduces a SabreDAV plugin that will block all older clients than 1.6.1 to connect and sync with the ownCloud instance.

This has multiple reasons:

1. Old ownCloud client versions before 1.6.0 are not properly working with sticky cookies for load balancers and thus generating sessions en masse
2. Old ownCloud client versions tend to be horrible buggy

In some cases we had in 80minutes about 10'000 sessions created by a single user. While this change set does not really "fix" the problem as 3rdparty legacy clients are affected as well, it is a good work-around and hopefully should force users to update their client

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/758e230c-4683-43b0-b6b6-15098d6096b8) was created.

LukasReschke commented:
In the future we need to come up with something that also works for regular third-party WebDAV clients… Creating tens of thousands of sessions for single users is not really what we should do. This is only a drop on the hot stone (though a big one…)

LukasReschke commented:
… please let's discuss the impact of this first before merging :) … 

also @MTRichards we really need to define which desktop versions are supported and which ones not :)

LukasReschke commented on lib/private/connector/sabre/blocklegacyclientplugin.php:
```diff
@@ -0,0 +1,58 @@
+		if(isset($versionMatches[1]) && version_compare($versionMatches[1], '1.6.0') === -1) {
+			// FIXME: Why can't we uset he ResponseInterface here?
+			http_response_code(503);
+			exit();
```
@evert Sorry for disturbing but I somewhat had problems to use the `ResponseInterface` here and then the `setStatus` stuff etc. – Do you have any idea what I am doing wrong here? Would be totally awesome if you could give me a hint :dancers: :beers:

LukasReschke replied:
(I also returned false)

DeepDiver1975 replied:
@LukasReschke I'd simple throw an exception - specifically ServiceUnavailable

evert replied:
Not sure why it didn't work, but @DeepDiver1975 is right, it's better to throw the exception there.

Also: `$request->getHeader('User-Agent')` instead of `$request->getRawServerValue('HTTP_USER_AGENT')`;

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/758e230c-4683-43b0-b6b6-15098d6096b8) completed: **21 new issues, 3 updated code elements**

MTRichards commented:
While I reserve the right to change this, or have exceptions based on real customer scenarios, my plan is to support the current release (1.8.x) and the previous major release (1.7.x). We would support any 1.7.x and 1.8.x, but that means we would need to force the update to the latest in that line if there is a problem. This also provides the opportunity for a customer to test the latest release before updating.

Once we ship the updater for on premise, and document how to host a linux package, there is no longer a good reason for not pushing upgrades. BUT, it is still up to the customer, and there are scenarios where this can't fully work.

So...Please don't just block older clients. This will lead to all sorts of unintended problems! It needs a graceful error message, or a notice of deprecation or something...

DeepDiver1975 commented on lib/private/connector/sabre/blocklegacyclientplugin.php:
```diff
@@ -0,0 +1,58 @@
+	function beforeHandler(RequestInterface $request) {
+		$userAgent = $request->getRawServerValue('HTTP_USER_AGENT');
+		preg_match("/(?:mirall\\/)([\d.]+)/i", $userAgent, $versionMatches);
+		if(isset($versionMatches[1]) && version_compare($versionMatches[1], '1.6.0') === -1) {
```
make this configurable?

DeepDiver1975 commented on lib/private/connector/sabre/blocklegacyclientplugin.php:
```diff
@@ -0,0 +1,58 @@
+	 */
+	function beforeHandler(RequestInterface $request) {
+		$userAgent = $request->getRawServerValue('HTTP_USER_AGENT');
+		preg_match("/(?:mirall\\/)([\d.]+)/i", $userAgent, $versionMatches);
```
@danimo is this enough to get all clients - even branded?

evert commented:
I also agree with @LukasReschke that sessions should ideally not ever be started =)

karlitschek commented:
I agree with @MTRichards We need to guide the users to update the client first before we block it completely.

DeepDiver1975 commented:
For customers we need this configurable (as in allow all, 1.6.0+ or 1.7.0 ....) - for the community we should by default disallow 1.6.0 and less - it is know that this client has issues and purely talking about the community: most recent clients are available in any linux distro or via OBS, Mac and Windows are already on our own update mechanism.

I'm all in for merging this into 8.1 with a config option.

@MTRichards @karlitschek does that make sense? THX

karlitschek commented:
i think it makes sense with a config option. what would be the default?
what do you think @mtrichards ?

MTRichards commented:
I agree - with a config option, a release note and some documentation, and a line in sample.config.php, A configurable value is acceptable. Clearly the versions allowed should be configurable (such as allow this version or newer, or deny this version or earlier where the admin enters the version), since we could see an oC client 2.4.1 accessing ownCloud 8.0.2 in the future.

That said, we can't expect our customers to know our version history numbering, so we need something better than enter something, such as comments in the config file that say the release date of the latest versions as of the time of release. And also in the documentation.

@carlaschroder adding release note tag for you.
@jnfrmarks for testing implications.

DeepDiver1975 commented:
> I also agree with @LukasReschke that sessions should ideally not ever be started =)



> That said, we can't expect our customers to know our version history numbering, so we need something better than enter something, such as comments in the config file that say the release date of the latest versions as of the time of release. And also in the documentation.

from a technical perspective we can only check on the explicit version number - the history is within the changelog of the client I assume. Would that be enough @MTRichards - or do I miss anything?

MTRichards commented:
I don't mind checking explicit version numbers in the code somewhere, but we can't expect an admin to know the explicit version numbers, or change a config option when a new client is released - they won't do it. 

An admin should just be able to say (and our default for new installations should be) the current version minus 2, so right now 1.7 and 1.8 clients would be allowed by default. 1.6 and earlier would be denied. Then also of course 1.9 or 2.0 or 2.1 etc would be allowed in this situation. In other words <=1.6 would be denied. 

All the admin has to do is know to leave the 1.6 in there, and all 1.6.x versions and earlier will be denied. When we release each ownCloud version, we just would need to increment the config file properly using autoconfig or the ee config file so that at the time of release, the latest 2 dot releases for clients can access.

Of course if we are upgrading customers, this should not be forced on our customers and users, rather it is optional.

DeepDiver1975 commented:
> An admin should just be able to say (and our default for new installations should be) the current version minus 2, so right now 1.7 and 1.8 clients would be allowed by default. 1.6 and earlier would be denied. Then also of course 1.9 or 2.0 or 2.1 etc would be allowed in this situation. In other words <=1.6 would be denied.

So with every ownCloud major release you want us to update the default config value for the blocked clients to be current version - 2 ?

jnfrmarks commented:
@rperezb 

FYI

DeepDiver1975 commented:
@MTRichards is there a need to have the same functionality for mobile clients?

THX @jnfrmarks for the hint :wink:

MTRichards commented:
@DeepDiver1975
>So with every ownCloud major release you want us to update the default config value for the blocked clients to be current version - 2 ?

Ideally, current major version - 2. 1.8 or 1.7 = major version.

That said, how deep down this hole do we want to get? It is possible that a problem in QT that we found makes 1.7.3 work, but not 1.7.2, 1.7.1 or 1.7.0. And 1.8.2, but not 1.8.1. So a customer would want to enable this for 1.7.2 and 1.8.2 and greater only.

The end solution might be 
1) A default version - 2. The best default is current major - 2
2) A config understanding two things: a specific version number and a blacklist.
The specific version number would be the version like 1.7.1. That would be the version cutoff, anything older (lower version) than that is denied. Then, a blacklist which denies specific versions in addition to this rule. 

For example, a setting such as 1.7.1, blacklist: 1.8.0, 1.8.1
This would be anything older than 1.7.1 denied. And also, 1.8.0 and 1.8.1 denied but otherwise anything 1.7.1 and newer is allowed.

For now, I would stick with 1) above and not worry about 2).

MTRichards commented:
@DeepDiver1975 
>@MTRichards is there a need to have the same functionality for mobile clients?

Probably. I know how we do propfinds in 8.1 changes, so older mobile versions will not work correctly. Asking @rperezb for more thoughts though.

The main difference though is that most people upgrade the mobile to the latest when they find they have a problem (or automatically), so the scale of the issue is a lot smaller. The problem really comes in older versions out there that can't upgrade, like iOS 5 devices. This is a really small %, but we should cover it anyway.
In this case, the default would be the latest version at the time of server release.

Same rule in my head: 
Android and iOS version number set. Any older version is denied, any equal or newer version is allowed. Same syntax, config file notation, release note and all.
@carlaschroder more documentation / release note fun!

DeepDiver1975 commented:
> Ideally, current major version - 2. 1.8 or 1.7 = major version.

I was not precise enough - I was talking about ownCloud Server major releases (as in 8.1, 8.2 ...).

DeepDiver1975 commented:
> And 1.8.2, but not 1.8.1. So a customer would want to enable this for 1.7.2 and 1.8.2 and greater only.

well - as of today: as soon as we are delivering e.g. 1.8.0 - there will be no further development in previous major versions. Based on this assumption we can keep the config entry quite simple.

```php

'minimum_desktop_client_version' => 1.6.2

dragotin commented:
I think to do that properly, you need to split the version into its three parts: 1.8.0 => major => 1, minor => 8 and patch => 0. With that, the implementation can support fine granular checks, such as
```
if( major_version >= 1 && minor_version >= 7 && patch_version > 1 ) {
  // allow client version 1.7.2 and 1.8.2 and bigger...
} else {
  // do not.
}
```
A lot of projects handle the versions according to that theme.

If the code checks for hard coded release strings, it might become tricky. I would not recommend that.

DeepDiver1975 commented:
@dragotin doing version comparison the right way is for sure necessary - this is out of question.
But we need a config parameter - which should be easy from my understanding

carlaschroder commented:
> @carlaschroder more documentation / release note fun!

Cool, I like fun :)

MTRichards commented:
@DeepDiver1975 
>Ideally, current major version - 2. 1.8 or 1.7 = major version.
>I was not precise enough - I was talking about ownCloud Server major releases (as in 8.1, 8.2 ...).

Now I am confused. How does this work with server version? I am missing something.

DeepDiver1975 commented:
> Now I am confused. How does this work with server version? I am missing something.

well - checking and rejecting a client is job of the server - we have to provide a default minimum client version - right?

MTRichards commented:
well - checking and rejecting a client is job of the server - we have to provide a default minimum client version - right? 

roger that. :ok: 
Perhaps I am not confused at all then. Also :ok:

dragotin commented:
@DeepDiver1975 yes, easy, but also functional! In this case I would do it right from the beginning on, rather than doing a "simple" solution that does not really fulfill the needs on the longer run. Just a stomach feeling...

LukasReschke commented:
There we go:

1. Added an `minimum.supported.desktop.version` switch in the configuration file, defaults to 1.7.0 for now.
2. Added unit tests
3. Switched to throw an exception

<hr/>
If we want to do this for mobile clients as well I can add a `minimum.supported.android.version` and `minimum.supported.ios.version` as well, in this case I require:

- [ ] The minimum supported version
- [ ] An user-agent that I can match on
- [ ] The verification that the mobile clients don't explode when they get a 503 returned

<hr/>

@DeepDiver1975 I changed the code to use an exception however every invalid client access will now be logged in the server log as exception, since the client does endless retries this gets quite spammy:

```json
{
    "reqId": "Sc2f1vM/let4+RTaX/0V",
    "remoteAddr": "::1",
    "app": "webdav",
    "message": "Exception: {\"Message\":\"Unsupported client version.\",\"Code\":0,\"Trace\":\"#0 [internal function]: OC\\\\Connector\\\\Sabre\\\\BlockLegacyClientPlugin->beforeHandler(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#1 \\/Users\\/lukasreschke\\/Documents\\/Programming\\/master\\/3rdparty\\/sabre\\/event\\/lib\\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\\n#2 \\/Users\\/lukasreschke\\/Documents\\/Programming\\/master\\/3rdparty\\/sabre\\/dav\\/lib\\/DAV\\/Server.php(456): Sabre\\\\Event\\\\EventEmitter->emit('beforeMethod', Array)\\n#3 \\/Users\\/lukasreschke\\/Documents\\/Programming\\/master\\/3rdparty\\/sabre\\/dav\\/lib\\/DAV\\/Server.php(254): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#4 \\/Users\\/lukasreschke\\/Documents\\/Programming\\/master\\/apps\\/files\\/appinfo\\/remote.php(81): Sabre\\\\DAV\\\\Server->exec()\\n#5 \\/Users\\/lukasreschke\\/Documents\\/Programming\\/master\\/remote.php(84): require_once('\\/Users\\/lukasres...')\\n#6 {main}\",\"File\":\"\\/Users\\/lukasreschke\\/Documents\\/Programming\\/master\\/lib\\/private\\/connector\\/sabre\\/blocklegacyclientplugin.php\",\"Line\":73}",
    "level": 4,
    "time": "2015-04-20T09:08:20+00:00"
}
```

Any ideas to prevent this?

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/d2030c2f-c5d3-403b-b24b-1984f2875260) was created.

LukasReschke committed:
Wording

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/4e0f1468-f093-44e5-b3c0-4c3582e69d0d) was created.

LukasReschke commented on config/config.sample.php:
```diff
@@ -799,6 +799,14 @@
+ * When adjusting this please be aware of the fact that older versions may
+ * be buggy and for best user experience we recommend to not change this value.
+ */
+'minimum.supported.desktop.version' => '1.7.0',
```
@carlaschroder @MTRichards for wording and check if that is okay so.

MTRichards replied:
* The minimum ownCloud desktop client version that is required to sync with
> + * this instance. Defaults to the official supported releases by ownCloud.
> + *
> + * When adjusting this please be aware of the fact that older versions may
> + * be buggy and for best user experience we recommend to not change this value.
 
Just a few minor adjustments.
 
The minimum ownCloud desktop client version that will be allowed to sync with this server instance. All connections made from earlier clients will be denied by the server. Defaults to the minimum officially supported ownCloud version at the time of release of this server version. 
 
When changing this, note that older unsupported versions of ownCloud may not function as expected, and could lead to permanent data loss for clients or other unexpected results.

LukasReschke replied:
Adjusted. Thanks!

LukasReschke commented on lib/private/connector/sabre/blocklegacyclientplugin.php:
```diff
@@ -0,0 +1,76 @@
+		preg_match("/(?:mirall\\/)([\d.]+)/i", $userAgent, $versionMatches);
+		if(isset($versionMatches[1]) &&
+			version_compare($versionMatches[1], $minimumSupportedDesktopVersion) === -1) {
+			throw new ServiceUnavailableException('Unsupported client version.');
```
Will spam the log as stated at https://github.com/owncloud/core/pull/15683#issuecomment-94401871

cc @DeepDiver1975

DeepDiver1975 replied:
with 503 the client will return forever - we need to silence it.

plain 500 - maybe???

DeepDiver1975 replied:
> with 503 the client will return forever - we need to silence it.
> 
> plain 500 - maybe???

@dragotin

LukasReschke commented:
That is what the client will show, maybe worth catching that 503 in the future or so?
![2015-04-20_11-18-41](https://cloud.githubusercontent.com/assets/878997/7227180/015a0b5e-e74f-11e4-815c-8f567db60eb2.png)

LukasReschke commented:
> I think to do that properly, you need to split the version into its three parts: 1.8.0 => major => 1, minor => 8 and patch => 0. With that, the implementation can support fine granular checks, such as

@dragotin That's what this code already does. `version_compare` will split version numbers this way.

dragotin commented:
Maybe 403 as @guruz suggests. But that needs to be tested with **OLD** clients before adding that to the server. Please find somebody who takes the responsibility for the tests.

DeepDiver1975 commented:
> Please find somebody who takes the responsibility for the tests.

@LukasReschke mind giving this a quick try? THX a lot!

LukasReschke commented:
> @LukasReschke mind giving this a quick try? THX a lot!

Last one I can test is 1.7.0 because apparently server disk space is sooooooo terribly expensive that we have to delete old builds all the time. Apparently we use a RAID-10 consisting of floppy disks or so…

If somebody gets me 1.6.0 I can test this as well… Mac or Linux versions are fine.

danimo commented:
Huh? There is a difference between the deletion (which was unfortunate and not coordiated) and the downloads mentioned on our websites (which we intentionally limit). the 1.6.x Mac clients are still there:

* https://download.owncloud.com/desktop/stable/ownCloud-1.6.0.201.dmg
* https://download.owncloud.com/desktop/stable/ownCloud-1.6.1.359.dmg
* https://download.owncloud.com/desktop/stable/ownCloud-1.6.2.551.dmg
* https://download.owncloud.com/desktop/stable/ownCloud-1.6.3.878.dmg
* https://download.owncloud.com/desktop/stable/ownCloud-1.6.4.1195.dmg

LukasReschke committed:
Use 403 instead a 50x response

LukasReschke commented:
403 seems to work on 1.6.4, 1.7.x and 1.8.0.

**1.6.4**
- [x] Set `minimum.supported.desktop.version` to `1.6.0`
- [x] Do an initial connection and setup of the client
- [x] Client is syncing
- [x] Set `minimum.supported.desktop.version` to `1.9.0`
- [x] Client shows error message.
- [x] Log is not spammed

**1.7.1**
- [x] Set `minimum.supported.desktop.version` to `1.7.0`
- [x] Do an initial connection and setup of the client
- [x] Client is syncing
- [x] Set `minimum.supported.desktop.version` to `1.9.0`
- [x] Client shows error message.
- [x] Log is not spammed

**1.8.0**
- [x] Set `minimum.supported.desktop.version` to `1.8.0`
- [x] Do an initial connection and setup of the client
- [x] Client is syncing
- [x] Set `minimum.supported.desktop.version` to `1.9.0`
- [x] Client shows error message.
- [x] Log is not spammed

<hr/>

Displayed to the user is something such as:
![2015-04-20_12-53-17](https://cloud.githubusercontent.com/assets/878997/7228709/39c09fbe-e75c-11e4-8721-e02c1b5f6dac.png)

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/943df19e-86e6-4a98-a5a8-5bbc04668525) was created.

LukasReschke commented:
@owncloud-bot Retest this please.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/943df19e-86e6-4a98-a5a8-5bbc04668525) completed: **6 new issues, 10 updated code elements**

MTRichards commented:
So it throws a 403 now?
What does the desktop client look like?

LukasReschke commented:
![2015-04-20 18_17_20](https://cloud.githubusercontent.com/assets/878997/7234746/8284c55e-e789-11e4-85ac-2ca520a3c31c.gif)

carlaschroder commented:
Release note: https://github.com/owncloud/documentation/commit/1ba4328253b4c9e1ee2e918f432f0b03752704f8

guruz commented:
@LukasReschke LOOOOOOOOOOOL.
@MTRichards :-D :-D :-D

DeepDiver1975 commented on tests/lib/connector/sabre/BlockLegacyClientPluginTest.php:
```diff
@@ -0,0 +1,118 @@
+	 */
+	public function oldDesktopClientProvider() {
+		return [
+			['Mozilla/5.0 (1.5.0) mirall/1.5.0'],
```
@danimo @dragotin are there more user agents we should take care of? csync???

LukasReschke committed:
Reword configuration text

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/766ff300-7808-411f-a2b3-bd1837fc147a) was created.

LukasReschke commented:
@owncloud-bot Retest this please.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/766ff300-7808-411f-a2b3-bd1837fc147a) completed: **7 new issues, 10 updated code elements**

DeepDiver1975 commented:
@LukasReschke please add this to the feature wiki - THX
@carlaschroder since you removed the 'release note' - has this been properly documented? THX

DeepDiver1975 commented:
> Release note: owncloud/documentation@1ba4328

okay - looks like - ability to read helps - THX @carlaschroder

DeepDiver1975 commented:
> @DeepDiver1975 I changed the code to use an exception however every invalid client access will now be logged in the server log as exception, since the client does endless retries this gets quite spammy:
> 
> {
>     "reqId": "Sc2f1vM/let4+RTaX/0V",
>     "remoteAddr": "::1",
>     "app": "webdav",
>     "message": "Exception: {\"Message\":\"Unsupported client version.\",\"Code\":0,\"Trace\":\"#0 [internal function]: OC\\\\Connector\\\\Sabre\\\\BlockLegacyClientPlugin->beforeHandler(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#1 \\/Users\\/lukasreschke\\/Documents\\/Programming\\/master\\/3rdparty\\/sabre\\/event\\/lib\\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\\n#2 \\/Users\\/lukasreschke\\/Documents\\/Programming\\/master\\/3rdparty\\/sabre\\/dav\\/lib\\/DAV\\/Server.php(456): Sabre\\\\Event\\\\EventEmitter->emit('beforeMethod', Array)\\n#3 \\/Users\\/lukasreschke\\/Documents\\/Programming\\/master\\/3rdparty\\/sabre\\/dav\\/lib\\/DAV\\/Server.php(254): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#4 \\/Users\\/lukasreschke\\/Documents\\/Programming\\/master\\/apps\\/files\\/appinfo\\/remote.php(81): Sabre\\\\DAV\\\\Server->exec()\\n#5 \\/Users\\/lukasreschke\\/Documents\\/Programming\\/master\\/remote.php(84): require_once('\\/Users\\/lukasres...')\\n#6 {main}\",\"File\":\"\\/Users\\/lukasreschke\\/Documents\\/Programming\\/master\\/lib\\/private\\/connector\\/sabre\\/blocklegacyclientplugin.php\",\"Line\":73}",
>     "level": 4,
>     "time": "2015-04-20T09:08:20+00:00"
> }
> 
> Any ideas to prevent this?

@LukasReschke this has been resolved from my understanding - right?

DeepDiver1975 commented:
ready to merge from my understanding :+1:

LukasReschke commented:
> @LukasReschke this has been resolved from my understanding - right?

Yes.

> @LukasReschke please add this to the feature wiki - THX

Done.

DeepDiver1975 commented on lib/private/connector/sabre/blocklegacyclientplugin.php:
```diff
@@ -0,0 +1,75 @@
+	 * @throws \Sabre\DAV\Exception\Forbidden If the client version is not supported
+	 */
+	public function beforeHandler(RequestInterface $request) {
+		$userAgent = $request->getHeader('User-Agent');
```
what happens if no user agent is defined?

LukasReschke replied:
Indeed. That will probably fail as `\Sabre\HTTP\MessageInterface::getHeader` returns `null` then. Let me add a unit test for this as well as a test. Nice catch :rocket:

LukasReschke replied:
https://github.com/owncloud/core/commit/ab9ea97d3af62fd1c9c1f813b84e5c45a585a8d8

DeepDiver1975 commented on lib/private/connector/sabre/blocklegacyclientplugin.php:
```diff
@@ -0,0 +1,75 @@
+	 */
+	public function beforeHandler(RequestInterface $request) {
+		$userAgent = $request->getHeader('User-Agent');
+		$minimumSupportedDesktopVersion = $this->config->getSystemValue('minimum.supported.desktop.version', '1.7.0');
```
what happens in case of miss configuration and the variable is set to:
- [ ] null
- [ ] false
- [ ] string

LukasReschke replied:
Well. :bomb: but honestly, if you add wrong data to your configuration ownCloud explodes in a lot of places :see_no_evil:

LukasReschke committed:
Catch not existing User-Agent header

In case of an not sent UA header consider the client as valid

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/51cb4a2b-510f-4cc9-b81f-7e456eb392f9) was created.

LukasReschke commented:
@owncloud-bot Retest this please.

DeepDiver1975 commented:
:shipit:

DeepDiver1975 commented:
one more review please @nickvergessen @th3fallen THX

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/51cb4a2b-510f-4cc9-b81f-7e456eb392f9) completed: **8 new issues, 8 updated code elements**

nickvergessen commented on tests/lib/connector/sabre/BlockLegacyClientPluginTest.php:
```diff
@@ -0,0 +1,129 @@
+<?php
```
lower case file names as per DeepDiver somewhere else in the PRs

LukasReschke replied:
Honestly, I will not change this on my own. If @DeepDiver1975 wants this then we shall document this properly and enforce this in Jenkins. We have a hell of test files that use upper case and PHPStorm will highlight PHPUnit files only using this syntax. Since we don't have that mad autoloader madness problem here it will also work quire well…

LukasReschke replied:
As in: Needs to get documented at https://doc.owncloud.org/server/8.0/developer_manual/general/codingguidelines.html :see_no_evil:

nickvergessen commented:
@DeepDiver1975 I don't see a comment on https://github.com/owncloud/core/pull/15683/files#r28717907 yet. Other then that, fine by me

carlaschroder commented:
doc updates https://github.com/owncloud/documentation/issues/1077
